### PR TITLE
Add support for PathLike classes & objects with __new__ but no __init__

### PR DESCRIPTION
**Pull request by JesseFarebro**
This solves most of the issues outlined in #36 and #37. Specifically,

1. Exactly like #37 it converts `os.PathLike` to `pathlib.Path`.
2. Additionally, if an object acts as `PathLike`, i.e., implements `os.PathLike` and is a type converter, i.e., `(arg: Union[str, Any]) -> T` then Tyro will parse this object as the "leaf node". This allows for custom `PathLike` objects like `epath.Path`.
3. Allows for recursing on objects that implement `__new__` and not `__init__`. One such object is `pathlib.Path` but there could be others out there.

Is there an argument to be made that any callable with a single positional argument that is `Union[str, Any]` should be treated as a leaf node? Or could there be a marker to specifically enable this functionality? If I remove the custom check for `PathLike` that stops the recursion you'd instead get something like `--path.args` instead of `--path` which I didn't think is desirable.

P.S. I changed `isinstance(typ, type)` -> `inspect.isclass(typ)` which I think makes things more clear. The actual implementation of `inspect.isclass` is `isinstance(typ, type)`.

### CODE REVIEW

File: tyro/_instantiators.py [hunk 1]
```diff
@@ -93,6 +102,47 @@ class UnsupportedTypeAnnotationError(Exception):
 )
 
 
+def _has_parsable_signature(typ: Union[Callable, TypeForm[Any]]) -> bool:
+    try:
+        inspect.signature(typ)
+    except Exception:
```
Comment (JesseFarebro) [inline tyro/_instantiators.py:108]: Maybe still only check for `ValueError`?

Comment (brentyi) [reply tyro/_instantiators.py:108]: What's the reason for broadening it to `Exception`?

Comment (JesseFarebro) [reply tyro/_instantiators.py]: No reason, just looked at the source for `inspect.signature` and it'll only raise `ValueError` so I changed this try block.

**Comment by JesseFarebro** [comment]
Rebased on main with #37.

**Review by brentyi — APPROVED** [review]
Going to double-check that this doesn't break some downstream dependents before releasing, but this looks good, thanks @JesseFarebro!

**Pull request merged**